### ISSUE TITLE
only-arrow-functions rule false-negative if function body contains 'this' keyword

### ISSUE THREAD

**Issue by Antony74** [palantir/tslint#4314]
### Bug Report

- __TSLint version__: 5.11
- __TypeScript version__: 3.1.6
- __Running TSLint via__: CLI/yarn

#### TypeScript code being linted

```ts
setTimeout(function() {
    this.oops();
}, 500);
```

with `tslint.json` configuration:

```json
{
    "defaultSeverity": "error",
    "extends": [
        "tslint:recommended"
    ],
    "jsRules": {},
    "rules": {
        "only-arrow-functions": true
    },
    "rulesDirectory": []
}
```

#### Actual behavior

No lint error.

![actual](https://user-images.githubusercontent.com/1577335/48917695-0b714f80-ee80-11e8-9968-bcaacc6c68a3.png)

#### Expected behavior

only-arrow-functions lint error

![expected](https://user-images.githubusercontent.com/1577335/48917724-41163880-ee80-11e8-8256-0a14869b8cfc.png)

It seems to be the presence of the 'this' keyword within the function body that causes this false-negative, which is unfortunate considering the rationale of the rule:

> Traditional functions don’t bind lexical scope, which can lead to unexpected behavior when accessing ‘this’.

**Comment by JoshuaKGoldberg** [comment]
Agreed, this is a bug. Looks like it was already fixed by https://github.com/palantir/tslint/pull/2229, which hasn't yet been released on npm.

You can work around this bug by supplying an explicit `this` type parameter: https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#specifying-the-type-of-this-for-functions

**Comment by Antony74** [comment]
Sorry, are you sure #2229 wasn't in 5.11?  If you scroll down to the bottom of this 5.11 source file...

https://github.com/palantir/tslint/blob/5.11.0/src/rules/onlyArrowFunctionsRule.ts

...you'll see the comment says "/** Generator functions and functions using `this` are allowed. */" which is the comment from the pull request.

Because this "feature" isn't in the [documentation for the rule](https://palantir.github.io/tslint/rules/only-arrow-functions/), I very reasonably reported the exception to this rule as a bug.  How do I disable the exception, please?  I don't want to see code like this:

```
class MyClass {

  hello() {
    console.log('hello');
  }

  myFunction() {
    setTimeout(function() {
      this.hello();
    }.bind(this), 500);
  }

}

```
It's much clearer as an arrow function because you don't need the bind.

**Comment by JoshuaKGoldberg** [comment]
Oh yes, good catch @Antony74 - that's in 5.11. Thanks!

If you'd like to send an issue mentioning the lack of documentation for this feature and/or a PR fixing it, that'd be great!

I tested this out locally using both your code snippets and saw the correct behavior. No complaints were logged when the functions had `this` in their bodies. Only a function without `this` gets logged:

```shell
C:\Code\tstest  (tstest@1.0.0)
λ tslint --version
5.11.0

C:\Code\tstest  (tstest@1.0.0)
λ type tslint.json
{
    "rules": {
        "only-arrow-functions": true
    }
}

C:\Code\tstest  (tstest@1.0.0)
λ type index.ts
class MyClass {
    myFunction() {
        setTimeout(function () {
            this.hello();
        }.bind(this), 500);
    }
}

setTimeout(function() {
    this.oops();
}, 500);

setTimeout(function () {}, 500);
C:\Code\tstest  (tstest@1.0.0)
λ tslint -c tslint.json index.ts

ERROR: index.ts:13:12 - non-arrow functions are forbidden


```

Are you sure you're running 5.11 locally? What does `tslint --version` print on your console?

**Comment by Antony74** [comment]
Sorry, I guess I'm still not making myself clear.  I agree the linter doesn't complain about those code snippets, and that this is correct behavior according to the feature request #2229, but I want those snippets to generate an error somehow, because they're not arrow functions and regardless of the 'this' keyword appearing in the body of the functions I want only-arrow-functions allowed.

This was a bug report, but I guess now it's a feature request for an additional option or something.

**Comment by Antony74** [comment]
Thanks @JoshuaKGoldberg I'm pleased to be coming over more clearly now. 

I'd have thought just the one option ("no-this-exception" maybe?) that basically undoes #2229 and complains about both your code snippets above.  I think it would be counter-productive to create a special case for `.bind(this)`, partly because of the added complexity, but mostly because the real danger of using a traditional function with `.bind(this)` is that one might accidently forget to add the `.bind(this)` on the end.

I don't mind having a look at this, if in principle you're prepared to accept a pull request for such a feature?

**Comment by JoshuaKGoldberg** [comment]
Let's hold off on a pull request - your enthusiasm is appreciated 😊 but we should wait on some more community feedback to see if this option is something a lot of folks would want. If it's just one or two people, it's probably not worth the added maintenance for TSLint. A good set of next steps would be:

1. File the feature request as a new issue describing your proposal and linking to this one
2. See if there's any demand for this and/or alternate proposals that would also fix your use case
3. If there is, TSLint could then accept a PR
